src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
+    f = invlink(m.f)
+    info = inv(hessian_function(x -> f(x), m.values.array[:, 1]))
+    return NamedArrays.NamedArray(info, (varnames, varnames))
+end
+
+StatsBase.coef(m::ModeResult) = m.values
+StatsBase.coefnames(m::ModeResult) = names(m.values)[1]
+StatsBase.params(m::ModeResult) = StatsBase.coefnames(m)
+StatsBase.vcov(m::ModeResult) = StatsBase.informationmatrix(m)
+StatsBase.loglikelihood(m::ModeResult) = m.lp
+
+####################
+# Optim.jl methods #
+####################
+
+"""
+    Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+
+Compute a maximum likelihood estimate of the `model`.
+
+# Examples
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.LikelihoodContext())
+    return optimize(model, OptimLogDensity(model, ctx, true), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+
+Compute a maximum a posterior estimate of the `model`.
+
+# Examples
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+map_est = optimize(model, MAP())
+
+# Use a different optimizer
+map_est = optimize(model, MAP(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.DefaultContext())
+    return optimize(model, OptimLogDensity(model, ctx, true), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+    # Do some initialization.
+    spl = DynamicPPL.SampleFromPrior()
+    init_params = model(f.vi, spl)
+    init_vals = f.vi[spl]
+
+    # Optimize!
+    M = Optim.optimize(f, init_vals, optimizer, args...; kwargs...)
+
+    # Get the VarInfo at the MLE/MAP point, and run the model to ensure 
+    # correct dimensionality.
+    f.vi[spl] = M.minimizer
+    invlink!(f.vi, spl)
+    vals = f.vi[spl]
+
+    # Make one transition to get the parameter names.
+    ts = [Turing.Inference.Transition(DynamicPPL.tonamedtuple(f.vi), DynamicPPL.getlogp(f.vi))]
+    varnames, _ = Turing.Inference._params_to_array(ts)
+
+    # Store the parameters and their names in an array.
+    vmat = NamedArrays.NamedArray(vals, varnames)
```

Comment:
Do you know how it compares to https://github.com/SciML/LabelledArrays.jl?

Reply:
I haven't tried `NamedArrays` but remember `LabelledArrays` being quite slow when it came to _constructing_ the array, but other than it seemed good. So in this case, `LabelledArrays` should be fine to use :+1:

Reply:
Potentially interesting though they don't have `LabelledArrays.jl`: https://github.com/mcabbott/AxisKeys.jl/blob/51550fe96f36937f7e4417adf77232f33f35657e/test/speed.jl

Reply:
The main reason I picked NamedArrays is because they have good `show` methods. LabelledArrays is pretty damn ugly and not good for presenting results.